system/lib/wasmfs/thread_utils.h

```diff
@@ -7,196 +7,45 @@
-  // etc., and also means we do not need to worry about the lifetime of the
-  // std::function in user code.
-  using Callback = std::function<void()>*;
-
-  //
-  // Run some work on thread. This is a synchronous (blocking) call. The thread
-  // where the work actually runs can do async work for us - all it needs to do
-  // is call the given callback function when it is done.
-  //
-  // Note that you need to call the callback even if you are not async, as the
-  // code here does not know if you are async or not. For example,
-  //
-  //  instance.invoke([](emscripten::SyncToAsync::Callback resume) {
-  //    std::cout << "Hello from sync C++ on the pthread\n";
-  //    (*resume)();
-  //  });
-  //
-  // In the async case, you would call resume() at some later time.
-  //
-  // It is safe to call this method from multiple threads, as it locks itself.
-  // That is, you can create an instance of this and call it from multiple
-  // threads freely.
-  //
-  void invoke(std::function<void(Callback)> newWork);
-
-  //==============================================================================
-  // End Public API
-
-private:
-  // The dedicated worker thread.
+// Helper class for synchronously proxying work to a dedicated worker thread,
+// including where the work is asynchronous.
+class ProxyWorker {
+  // The queue we use to proxy work and the dedicated worker.
+  ProxyingQueue queue;
   std::thread thread;
 
-  // Condition variable used for bidirectional communication between the worker
-  // thread and invoking threads.
-  std::condition_variable condition;
-  std::mutex mutex;
-
-  // The current state of the worker thread. New work can only be submitted when
-  // in the `Waiting` state.
-  enum State {
-    Waiting,
-    WorkAvailable,
-    ShouldExit,
-  } state = Waiting;
-
-  // Increment the count every time work is finished. This will allow invokers
-  // to detect that their particular work has been completed even if some other
-  // invoker wins the race and submits new work before the original invoker can
-  // check for completion.
-  std::atomic<uint32_t> workCount{0};
-
-  // The work that the dedicated worker thread should perform and the callback
-  // that needs to be called when the work is finished.
-  std::function<void(Callback)> work;
-  std::function<void()> resume;
-
-  static void* threadMain(void* arg) {
-    // Schedule ourselves to start processing incoming work requests.
-    emscripten_async_call(threadIter, arg, 0);
-    return 0;
+public:
+  // Spawn the worker thread.
+  ProxyWorker()
+    : queue(), thread([&]() {
+        // Process the queue in case any work came in while we were starting up,
+        // then sit in the event loop performing work as it comes in.
+        queue.execute();
+        emscripten_exit_with_live_runtime();
+      }) {}
+
+  // Kill the worker thread.
+  ~ProxyWorker() {
+    pthread_cancel(thread.native_handle());
```

Comment:
I would prefer to avoid using `pthread_cancel` if at all possible. 

Can we come up with standard mechanism for sending some kind shutdown message to the queue?

Reply:
One option would be to have the worker thread wait for work inside a loop using a traditional condition variable. Then it could exit the loop and shut itself down normally once we signal it to do so. But it's kind of nice to be able to have it so simply wait in the JS event loop instead. Given the need to tell another thread sitting in its event loop to shut down, wouldn't any mechanism for doing so basically be identical to what `pthread_cancel` does in that case?